### PR TITLE
Dev Container improvements

### LINKED ISSUE #14292: Improve handling of dev container workspaces
### Bug Description:
There are no designated dev-container entries (e.g., `[dev-container] <workspaceName>`) in the recent workspaces list that would allow for directly reopening a workspace as a development container. Currently, reopening a dev container requires opening the folder as a normal workspace first, then manually reopening it as a dev container, which could be streamlined into a single click.

Additionally, when closing Theia with a development container as the last active workspace, the expected behavior is for Theia to reopen in the same dev container upon restart. However, Theia currently reopens the workspace in its normal state, requiring manual reopening as a dev container, adding unnecessary steps to the workflow.

Furthermore, there is no clear indication in the workspace name (displayed at the top of the application) to show that a workspace is opened as a dev container. The only visual cue is in the status bar, which can be easily overlooked.

Improving the handling of development container workspaces by:
- [ ] Adding designated entries in the recent workspaces list for direct reopening as dev containers.
- [ ] Ensuring Theia reopens in the last active dev container automatically after restart.
- [ ] Providing clearer visual indicators (such as in the workspace name) when a workspace is opened as a dev container.

### Steps to Reproduce:
1. Open a workspace as a dev container.
2. Observe that the workspace name does not indicate that it is a dev container.
3. Close Theia.
4. Reopen Theia.
5. Notice that the workspace is no longer in the dev container.
6. Check the recent workspaces list and observe that there is no entry to directly reopen the workspace as a dev container.

### Additional Information:
- Operating System: Ubuntu 22.04
- Theia Version: Theia IDE v1.54.0

### PR DESCRIPTION
**Pull request by sgraband**
#### What it does

Contributes to https://github.com/eclipse-theia/theia/issues/14294

This PR adds several new features, bug fixes, and quality improvements to the dev container support in Theia:

**New features:**
- **Variable resolver fixes**: Handle variables embedded in strings in devcontainer.json (e.g. `${localEnv:HOME}/path`)
- **MountsContribution fix**: Pass through all mount options (source, target, type) correctly instead of silently dropping them
- **SettingsContribution fix**: Fix JSON quoting when injecting preferences through the shell by base64-encoding complex objects
- **portsAttributes support**: Implement `portsAttributes` from devcontainer.json, allowing port labels, protocols, and auto-forward behavior to be configured
- **Attach to Running Container**: Add a command to attach to an already-running Docker container without a devcontainer.json
- **SSH credential injection, git identity sharing, and shell detection**: Automatically set up an isolated SSH directory with key forwarding, inject host `.gitconfig` (with SSH signing key rewriting), configure a login shell with SSH agent, and share host credentials into the container
- **Auto-reopen in last active dev container after restart**: On application restart, automatically reconnect to the previously used dev container
- **Suggest to reopen workspace in dev container**: When opening a workspace that contains a devcontainer.json, show a notification suggesting to reopen in a container
- **Rebuild Container command**: Add a command to rebuild the dev container from scratch (removing the old container) while preserving the devcontainer file context
- **Ports view and port labels**: Expose the Ports view in the remote package and display port labels from `portsAttributes`

#### How to test

Prerequisites: Docker must be installed and running.

1. **Reopen in Container**
   - Open a workspace that contains a `.devcontainer/devcontainer.json`
   - Run `Dev Container: Reopen in Container` from the command palette
   - Verify Theia connects to the container and the workspace is available

2. **Suggestion notification**
   - Open a workspace with a devcontainer.json from a local (non-remote) session
   - Verify a notification appears suggesting to reopen in a container
   - Click "Reopen in Container" and verify it works

3. **Rebuild Container**
   - While connected to a dev container, run `Dev Container: Rebuild Container`
   - Verify the old container is removed and a new one is created
   - Verify the workspace reconnects automatically

4. **Attach to Running Container**
   - Start a Docker container manually (e.g. `docker run -it ubuntu bash`)
   - Run `Dev Container: Attach to Running Container`
   - Select the running container and verify Theia connects

5. **SSH credentials and git identity**
   - Ensure you have `~/.ssh/config` and `~/.gitconfig` on the host
   - Reopen in a container and verify:
     - `git config user.name` and `user.email` are inherited
     - `ssh -T git@github.com` works (key forwarding)
     - SSH agent is running (`echo $SSH_AUTH_SOCK`)

6. **Port forwarding with labels**
   - Use a devcontainer.json with `forwardPorts` and `portsAttributes` (e.g. label, protocol)
   - Verify ports are forwarded and labels appear in the Ports view

7. **Variable resolver with embedded variables**
   - Use a devcontainer.json with variables embedded in strings, e.g.:
     ```json
     { "remoteEnv": { "MY_VAR": "${localEnv:HOME}/projects" } }
     ```
   - Reopen in the container and verify `echo $MY_VAR` resolves to the host's `$HOME/projects` (not the literal `${localEnv:HOME}` string)

8. **Mount options pass-through**
   - Add a `mounts` entry in devcontainer.json using the string form with extra options, e.g.:
     ```json
     { "mounts": ["type=bind,source=${localEnv:HOME}/.config,target=/home/vscode/.config,readonly"] }
     ```
   - Reopen in the container and verify the mount exists (`mount | grep .config`), is read-only, and uses the correct source/target paths

9. **Settings with complex JSON values**
   - Add a complex (object/array) setting in devcontainer.json, e.g.:
     ```json
     { "settings": { "editor.tokenColorCustomizations": { "comments": "#FF0000" } } }
     ```
   - Reopen in the container and verify the preference is applied correctly in the running Theia instance (open Settings and search for the key)
   - Previously, object values were broken by shell quoting when passed as `--set-preference` CLI arguments

10. **Auto-reopen after restart**
    - Connect to a dev container, then close and reopen Theia
    - Verify it automatically reconnects to the same container

#### Breaking changes

- [ ] This PR introduces breaking changes and requires careful review. If yes, the breaking changes section in the [changelog](https://github.com/eclipse-theia/theia/blob/master/CHANGELOG.md) has been updated.

#### Review checklist

- [x] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)
- [x] User-facing text is internationalized using the `nls` service (for details, please see the [Internationalization/Localization section](https://github.com/theia-ide/theia/blob/master/doc/coding-guidelines.md#internationalizationlocalization) in the Coding Guidelines)

#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

File: packages/workspace/src/browser/workspace-service.ts [hunk 1]
```diff
@@ -121,8 +121,48 @@ export class WorkspaceService implements FrontendApplicationContribution, Worksp
...
+        if (wsUriString) {
+            const wsUri = new URI(wsUriString);
+            if (wsUri.scheme !== 'file') {
+                let handled = false;
+                for (const handler of this.openHandlerContribution.getContributions()) {
+                    if (await handler.canHandle(wsUri)) {
+                        try {
+                            // openWorkspace reloads the window on success, so if we
+                            // reach the timeout the connection attempt is hanging.
+                            await Promise.race([
```
Comment (ndoschek) [inline packages/workspace/src/browser/workspace-service.ts:134]: The 15-second timeout may be too short for dev container reconnection. `openWorkspace()` calls `doOpenInContainer()` → `connectToContainer()`, which can pull Docker images, create containers, and set up remote connections which can easily exceed 15 seconds on first run or slow networks. When the timeout fires, it falls back to a local workspace while the container creation continues in the background, leading to a confusing state. Consider increasing the timeout or making it configurable.

File: packages/dev-container/src/electron-node/devcontainer-contributions/cli-enhancing-creation-contributions.ts [hunk 2]
```diff
@@ -59,7 +59,13 @@ export class SettingsContribution implements RemoteCliContribution, ContainerCre
             ...(this.currentConfig.customizations?.vscode?.settings ?? [])
         };
         this.currentConfig = undefined;
-        return Object.entries(settings).map(([key, value]) => `--set-preference=${key}=${JSON.stringify(value)}`) ?? [];
+        return Object.entries(settings).map(([key, value]) => {
+            const jsonValue = JSON.stringify(value);
+            if (value && typeof value === 'object') {
+                return `--set-preference=${key}=base64:${Buffer.from(jsonValue).toString('base64')}`;
```
Comment (ndoschek) [inline packages/dev-container/src/electron-node/devcontainer-contributions/cli-enhancing-creation-contributions.ts:65]: There are no tests for the base64 encode/decode round-trip between `SettingsContribution.enhanceArgs()` and `PreferenceCliContribution.setArguments()`. Since this is a serialization boundary between two separate processes, consider adding unit tests covering edge cases: nested objects, arrays, values containing `=` signs, and empty objects.

File: packages/dev-container/src/electron-node/devcontainer-contributions/main-container-creation-contributions.ts [hunk 3]
```diff
@@ -191,6 +226,262 @@ export class ContainerEnvContribution implements ContainerCreationContribution {
...
+                        }
+                    }
+                }
+            }
+
+            // Generate a dedicated ed25519 keypair if none exists
+            const keyPath = path.join(isolatedDir, 'id_ed25519');
+            if (!await fs.pathExists(keyPath)) {
+                await new Promise<void>((resolve, reject) => {
+                    cp.exec(`ssh-keygen -t ed25519 -f "${keyPath}" -N "" -C "theia-dev-container"`, err => {
```
Comment (ndoschek) [inline packages/dev-container/src/electron-node/devcontainer-contributions/main-container-creation-contributions.ts:409]: `ensureIsolatedSshDir()` silently generates a new ed25519 keypair at `~/.theia/dev-container/ssh/id_ed25519` if none exists. While it logs a message asking the user to register the public key, auto-generating cryptographic material on the user's machine could be surprising. Consider making this opt-in or using a more prominent notification (e.g., via `MessageService`).

File: packages/dev-container/src/electron-browser/dev-container-suggestion-contribution.ts [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        }
+
+        try {
+            const devcontainerFiles = await this.connectionProvider.getDevContainerFiles(workspace.resource.path.toString());
+            if (devcontainerFiles.length > 0) {
+                const reopenAction = nls.localize('theia/remote/dev-container/reopenInContainer', 'Reopen in Container');
+                const dontShowAgain = nls.localizeByDefault("Don't Show Again");
+                const result = await this.messageService.info(
+                    nls.localize('theia/remote/dev-container/suggestion',
+                        'This workspace has a dev container configuration. Would you like to reopen it in a container?'),
```
Comment (ndoschek) [inline packages/dev-container/src/electron-browser/dev-container-suggestion-contribution.ts:79]: The 'Don't Show Again' button is captured but never persisted. Clicking it dismisses the dialog, but the notification will reappear on next startup because there's no `storageService.setData()` or preference write to remember the user's choice.
Would also be fine to drop this option for now and handle in a follow up.

File: packages/dev-container/src/electron-browser/container-connection-contribution.ts [hunk 5]
```diff
@@ -110,26 +187,115 @@ export class ContainerConnectionContribution extends AbstractRemoteRegistryContr
...
+            text: nls.localize('theia/remote/dev-container/rebuilding', 'Rebuilding dev container...')
+        });
+
+        try {
+            // When inside a remote container, read the stored context instead of
+            // scanning the filesystem (the RPC goes to the local backend which
+            // doesn't have the container's workspace path).
+            const ctx = await this.storageService.getData<DevContainerContext | undefined>(ACTIVE_DEV_CONTAINER_CONTEXT);
+            if (ctx) {
+                progress.report({ message: 'Removing old container...' });
```
Comment (ndoschek) [inline packages/dev-container/src/electron-browser/container-connection-contribution.ts]: The user facing progress and error messages are not localized

File: packages/dev-container/src/electron-node/devcontainer-contributions/main-container-creation-contributions.ts [hunk 6]
```diff
@@ -191,6 +226,262 @@ export class ContainerEnvContribution implements ContainerCreationContribution {
...
+    @inject(ILogger)
+    protected readonly logger: ILogger;
+
+    async handleContainerCreation(createOptions: Docker.ContainerCreateOptions, containerConfig: DevContainerConfiguration): Promise<void> {
+        const mounts = createOptions.HostConfig?.Mounts ?? [];
+        const hasExistingMount = (targetSuffix: string): boolean =>
+            mounts.some(m => m.Target?.endsWith(targetSuffix));
+
+        // SSH: bind-mount an isolated SSH directory instead of the real ~/.ssh
+        const sshSigningInfo = await this.detectSshSigningKey();
```
Comment (ndoschek) [inline packages/dev-container/src/electron-node/devcontainer-contributions/main-container-creation-contributions.ts:243]: `detectSshSigningKey()` is called twice: once here in `handleContainerCreation` and again in `handlePostCreate` (line 287). Each call runs `git config --global --get` via `child_process.exec`. Consider caching the result from the first call (e.g., as an instance field) to avoid redundant process spawning and potential inconsistency if git config changes between calls.

File: packages/dev-container/src/electron-node/devcontainer-contributions/main-container-creation-contributions.ts [hunk 7]
```diff
@@ -191,6 +226,262 @@ export class ContainerEnvContribution implements ContainerCreationContribution {
...
+        // env vars must be baked into the container environment.
+        if (!createOptions.Env) {
+            createOptions.Env = [];
+        }
+        const setIfMissing = (key: string, value: string): void => {
+            if (!createOptions.Env!.some(e => e.startsWith(key + '='))) {
+                createOptions.Env!.push(`${key}=${value}`);
+            }
+        };
+        setIfMissing('THEIA_SHELL', '/bin/bash');
```
Comment (ndoschek) [inline packages/dev-container/src/electron-node/devcontainer-contributions/main-container-creation-contributions.ts]: Hardcodes `/bin/bash` as the default shell. Some container images (e.g., Alpine-based) don't include bash. If `/bin/bash` doesn't exist, terminal sessions will fail to start. Consider detecting the available shell inside the container (e.g., `which bash || which sh`) in `handlePostCreate`, or at least falling back to `/bin/sh`.

File: packages/dev-container/src/electron-node/remote-container-connection-provider.ts [hunk 8]
```diff
@@ -182,6 +182,103 @@ export class DevContainerConnectionProvider implements RemoteContainerConnection
...
+            const report: RemoteStatusReport = message => progress.report({ message });
+            report('Connecting to remote system...');
+
+            const remote = new RemoteDockerContainerConnection({
+                id: generateUuid(),
+                name: containerInfo.Name.replace(/^\//, ''),
+                type: 'Dev Container',
+                docker,
+                container,
+                config: {} as DevContainerConfiguration,
```
Comment (ndoschek) [inline packages/dev-container/src/electron-node/remote-container-connection-provider.ts]: `attachToContainer()` uses `config: {} as DevContainerConfiguration` which bypasses type safety. Downstream code (e.g., `shutdownContainer()`) accesses `config.shutdownAction` and `config.dockerComposeFile` — both will be `undefined` on the empty object. Currently this works via `??` / `?` operators, but it's fragile: if new required config properties are added, this could break silently. Consider creating a `DevContainerConfiguration.empty()` factory or using `Partial<DevContainerConfiguration>`.

File: packages/dev-container/src/electron-node/remote-container-connection-provider.ts [hunk 9]
```diff
@@ -265,7 +362,7 @@ export class RemoteDockerContainerConnection implements RemoteConnection {
             socket.pipe(stream);
             ttySession.modem.demuxStream(stream, socket, socket);
         } catch (e) {
-            console.error(e);
+            this.logger.error(e);
```
Comment (ndoschek) [inline packages/dev-container/src/electron-node/remote-container-connection-provider.ts]: Not sure about this, I think these are unnecessary changes, we should keep console: https://github.com/eclipse-theia/theia/blob/master/doc/coding-guidelines.md#logging

File: packages/workspace/src/browser/workspace-service.ts [hunk 10]
```diff
@@ -373,7 +415,22 @@ export class WorkspaceService implements FrontendApplicationContribution, Worksp
     }
 
     protected setMostRecentlyUsedWorkspace(): void {
-        this.server.setMostRecentlyUsedWorkspace(this._workspace ? this._workspace.resource.toString() : '');
+        if (!this.isRemoteSession()) {
```
Comment (ndoschek) [inline packages/workspace/src/browser/workspace-service.ts:418]: The `isRemoteSession()` guard was correctly added to `setWorkspace()` and `setMostRecentlyUsedWorkspace()` (on stop), but `close()` (line 600) still unconditionally calls `server.setMostRecentlyUsedWorkspace('')`. Similarly, `save()` (line 723) writes MRU unconditionally. In a remote session, calling `close()` would overwrite the `devcontainer://` MRU URI with an empty string, breaking auto-reopen on restart. Consider adding the same `isRemoteSession()` guard there.

**Review by ndoschek — COMMENTED** [review]
Thanks a lot for this PR @sgraband!
I tested it and things work well on my end 🎉 
I've added a few inline comments, please have a look.

One bit of housekeeping: the commits reference several issues this PR fixes/contributes to, but they're not linked from the PR description yet. Could you add/link them?

On the bigger picture: a natural next step would be to make "open folder in dev container" smoother by having Theia determine the correct workspace folder automatically from the devcontainer metadata (`workspaceFolder`, mounts...) instead of relying on the user to navigate there. But of course fine as a follow-up.

File: packages/workspace/src/browser/workspace-service.ts (hunk 1 above)
Comment (sgraband) [reply packages/workspace/src/browser/workspace-service.ts:134]: I feel like increasing the timeout to 2 minutes should be fine. Without configuration (for now). This should only really be relevant, when the user deleted the docker image and then restarts Theia and the dev container is the most recent workspace if i am not missing something. So this does not really happen that often i believe, therefore a setting is not really necessary imho.

File: packages/dev-container/src/electron-node/remote-container-connection-provider.ts (hunk 9 above)
Comment (ndoschek) [reply packages/dev-container/src/electron-node/remote-container-connection-provider.ts]: Looks like a leftover, but we should remove this here too.

File: packages/dev-container/src/electron-node/docker-container-service.ts [hunk 11]
```diff
@@ -81,7 +84,7 @@ export class DockerContainerService {
                 }
             } catch (e) {
                 container = undefined;
-                console.warn('DevContainer: could not find last used container');
+                this.logger.warn('DevContainer: could not find last used container');
```
Comment (ndoschek) [inline packages/dev-container/src/electron-node/docker-container-service.ts]: Same here, we can omit this change

File: packages/dev-container/src/electron-browser/container-connection-contribution.ts [hunk 12]
```diff
@@ -110,26 +187,115 @@ export class ContainerConnectionContribution extends AbstractRemoteRegistryContr
...
+        this.containerOutputProvider.openChannel();
+
+        const connectionResult = await this.connectionProvider.attachToContainer(selected.container.id);
+        this.openRemote(connectionResult.port, false, connectionResult.workspacePath);
+    }
+
+    async rebuildContainer(): Promise<void> {
+        this.containerOutputProvider.openChannel();
+        const progress = await this.messageService.showProgress({
+            text: nls.localize('theia/remote/dev-container/rebuilding', 'Rebuilding dev container...')
```
Comment (ndoschek) [inline packages/dev-container/src/electron-browser/container-connection-contribution.ts]: As this is basically the prefix of the progress monitor, we should drop the ... to avoid reporting like this: `Rebuilding dev container...: Removing old container...`

```suggestion
            text: nls.localize('theia/remote/dev-container/rebuilding', 'Rebuilding dev container')
```

File: packages/dev-container/src/electron-node/devcontainer-contributions/cli-enhancing-creation-contributions.spec.ts [hunk 13]
```diff
@@ -0,0 +1,519 @@
+// *****************************************************************************
+// Copyright (C) 2025 EclipseSource GmbH.
```
Comment (ndoschek) [inline packages/dev-container/src/electron-node/devcontainer-contributions/cli-enhancing-creation-contributions.spec.ts]: Should be 2026

**Review by ndoschek — COMMENTED** [review]
Thank you for the updates @sgraband!
Overall looks great to me, just a few minor comments inline.

I tested the mentioned features and almost all features work well on my end.

As briefly mentioned offline, the `remoteEnv` do not work for me at all, independent if they contain a variable to resolve, and this is also the case on current master already. Perhaps this is a setup issue on my end? Do they work reliably for you?

And finally, I also noticed, that settings defined via the devcontainer.json or passed via the cli argument end up written in the user settings file, which I would not expect tbh. But we can do this in a follow up as this was not introduced with this PR.

**Review by ndoschek — APPROVED** [review]
Thanks, LGTM 👍
